lib/eval.ml

```diff
@@ -364,77 +367,85 @@ module MakeInterpreter (T : Target) = struct
-                      | Int {width} -> Expression.Int (Info.dummy, {value = num; width_signed = Some (width,true)})
-                      | Bit {width} -> Expression.Int (Info.dummy, {value = num; width_signed = Some (width,false)})
-                      | Bool -> Expression.Int (Info.dummy, {value = num; width_signed = None})
-                      | _ -> failwith "unsupported type" in
-        let typed_exp : Expression.typed_t = {expr = pre_exp; typ = t; dir = Directionless} in
-        let exp = (Info.dummy, typed_exp) in
-        if String.contains s '*'
-        then begin
-        let pre_exp' = Expression.Mask {expr = exp; mask = exp} in
-        let typed_exp' : Expression.typed_t = {expr = pre_exp'; typ = Void; dir = Directionless} in
-        Some (Info.dummy, typed_exp') end
-        else Some exp in
+    match s with
+    | None -> None
+    | Some (s, t) ->
+      let num = s |> replace_wildcard |> int_of_string |> Bigint.of_int in
+      let rec pre_expr_of_typ env (t : Type.t) =
+        match t with
+        | Integer -> Expression.Int (Info.dummy, {value = num; width_signed = None})
+        | Int {width} -> Expression.Int (Info.dummy, {value = num; width_signed = Some (width,true)})
+        | Bit {width} -> Expression.Int (Info.dummy, {value = num; width_signed = Some (width,false)})
+        | Bool -> Expression.Int (Info.dummy, {value = num; width_signed = None})
+        | TypeName n -> pre_expr_of_typ env (EvalEnv.find_typ n env)
+        | _ -> failwith "unsupported type" in
+      let pre_exp = pre_expr_of_typ env t in
+      let typed_exp : Expression.typed_t = {expr = pre_exp; typ = t; dir = Directionless} in
+      let exp = (Info.dummy, typed_exp) in
+      if String.contains s '*'
+      then begin
+      let pre_exp' = Expression.Mask {expr = exp; mask = exp} in
+      let typed_exp' : Expression.typed_t = {expr = pre_exp'; typ = Void; dir = Directionless} in
+      Some (Info.dummy, typed_exp') end
+      else Some exp
+  
+  and convert_action env actions (name, args) =
+      let action_name' = Types.BareName (Info.dummy, name) in
+      (*let action_type = EvalEnv.find_typ action_name' env in*)
+      let type_params = EvalEnv.find_decl action_name' env |> assert_action_decl in
+      let existing_args = List.fold_left actions
+                          ~f:(fun acc a -> if Types.name_eq (snd a).action.name action_name'
+                                           then (snd a).action.args
+                                           else acc)
+                          ~init:[] in
+      let ctrl_args = match_params_to_args type_params args |> List.map ~f:(convert_expression env) in
+      let pre_action_ref : Table.pre_action_ref =
+        { annotations = [];
+          name = action_name';
+          args = existing_args @ ctrl_args } in
+      let action : Table.typed_action_ref = { action = pre_action_ref; typ = Void } in (*type is a hack*)
```

Comment:
We should have a validator/typechecker that turns `Ast` stuff into regular match/entry data using the typechecker environment but I think that might require some changes to the typechecker to make it easy to get at the types of actions that are nested inside a control etc.

Reply:
thats a good idea should we open a separate issue?

Reply:
Yeah I'll do that right now

Reply:
#232